### DISCUSSION TITLE
Why do I get a "push" event instead of a "pull_request" on a PR merge to master

### DISCUSSION THREAD

**Discussion by Grraahaam** [community/community#26558] [Actions]
<p><strong>Hello folks!</strong></p>
<p>The problematic (may be an issue/bug) : When I merge a PR from a branch to <code>master</code> (<strong>default</strong>) or synch a branch from <code>master</code>, in my action workflow, I’m getting a “push” event, but I’m expecting the “pull_request” one</p>
<p>For you to get an overview of the git structure flow I’m using, here it is :</p>
<ul>
<li>We have 3 branches, <code>develop</code>, <code>staging</code> and <code>master</code> (which is the default branch)</li>
<li>The merge flow is <code>develop</code> → <code>staging</code> → <code>master</code>
</li>
<li>CI/CD triggers on <strong>closed</strong> and <strong>merged</strong> PR from <code>develop</code> → <code>staging</code>, and also when PUSH on <code>master</code> (hotfix scenario), condition below :</li>
</ul>
<pre><code class="lang-yaml">job-name:
  if: (github.event.pull_request.merged == true &amp;&amp; github.event.pull_request.head.ref == 'develop') || github.event_name == 'push'
</code></pre>
<ul>
<li>CI/CD workflow automatically build docker images if files within a given directory are edited</li>
</ul>
<p>So far, PRs from <code>develop</code> to <code>staging</code> (and reverse for sync) trigger the expected event and workflow job, I’m getting <code>github.event_name : "pull_request"</code> and <code>github.head_ref : "develop"</code> or <code>"staging"</code>, depends if its a sync or not", which I use in my job condition (<code>if</code>)</p>
<p>But when it comes to PRs from <code>staging</code> to <code>master</code>, I’m expecting the same event (with <code>github.event_name : "pull_request"</code> and <code>github.head_ref : "staging"</code>) but I’m getting “push” instead, which triggers the action also on master branch, which breaks my CI/CD control structure. It shouldn’t trigger, according to the condition (<code>if</code>) above and the expected event.</p>
<p>And “push” events doesn’t seem to include any value I can use to detect whether it comes from <code>staging</code> or <code>master</code> itself, except for the PR commit name but I don’t want to rely on names which can change over time.</p>
<p>Is every “pull_request” event followed by a “push” one? Because I’ve been trying the following and got two events, so two running workflow when PR staging to master :</p>
<pre><code class="lang-yaml">on:
  pull_request:
    branches:
      - 'staging'
      - 'master'
    types:
      - closed
      - synchronize # Not used yet, but could be useful for pre-test checks
    paths:
      - '.github/workflows/generic-docker.yaml'
      - 'service-one/**'
      - 'service-two/**'
  # Hotfix process when pushing directly to master
  push:
    branches:
      - 'master'
</code></pre>
<p>Am I misunderstanding some points here, I’ve been searching documentation on this, but still no luck…</p>
<p>The workflow file :</p>
<pre><code class="lang-yaml">on:
  # Clean process flow when merging branches
  pull_request:
    branches:
      - 'staging'
    types:
      - closed
      - synchronize # Not used yet, but could be useful for pre-test checks
    paths:
      - '.github/workflows/generic-docker.yaml'
      - 'service-one/**'
      - 'service-two/**'
  # Hotfix process when pushing directly to master
  push:
    branches:
      - 'master'
    paths:
      - '.github/workflows/generic-docker.yaml'
      - 'service-one/**'
      - 'service-two/**'

jobs:
  generic-docker:
    runs-on: ubuntu-latest
    if: (github.event.pull_request.merged == true &amp;&amp; github.event.pull_request.head.ref == 'develop') || github.event_name == 'push'
    steps:
    	# Run specific steps depending on which files changed
    	# ...
</code></pre>
<p>Does anyone have a clue on how I could solve this, in a reliable way? Thanks for reading!</p>
<p>Thanks!</p>

**Comment by melink14** [comment, marked as answer]
<p>Just commenting briefly to say what you’re experiencing is expected. <code>pull_request</code> event triggers when a PR event occurs targeting the branches specified in the workflow. <code>push</code> event is triggered whenever a commit is pushed to the specified branches.</p>
<p>Since a PR merge is just a commit pushed to the target branch it will trigger <code>push</code> events as you’ve seen.</p>
<p>I can’t think of an easy way to accomplish what you want but one more complicated possibility would be to make an API call to see if the push commit is seen in any recent pull requests. If it’s not then you know it’s a hotfix push.</p>
<p>edit: clarified that your events don’t happen on open. Also, in the end you’ll probably want a workflow file similar to the one you tried where you list both <code>staging</code> and <code>master</code> in the <code>pull_request</code> section.</p>

**Comment by Grraahaam** [comment]
<p>Hey <a class="mention" href="/melink14">@melink14</a>! Thanks for the quick reply!</p>
<p>You just made my day :sun_with_face: I do now understand a bit more, is there any documentation about those behaviors? I already did search a lot, I may not be good at searching :upside_down_face:?</p>
<p>Anyway, I’m just answering briefly, I’m testing my new CI/CD workflow and it seems to work as expected! TL;DR; I’m filtering hotfixes on the base branch’s name, and instead of a <code>push</code> event I’m only using the <code>pull_request</code> one. And hotfixes need to have their own branches, from the base branch they’re aimed to fix.</p>
<p>This basically did the trick:</p>
<pre><code class="lang-yaml">name: Generic docker
on:
  pull_request:
    branches:
      - 'staging'
      - 'master'
    types:
      - closed
      - synchronize
    paths:
      - '.github/workflows/generic-docker.yaml'
      - 'service-one/**'
      - 'service-two/**'
jobs:
  generic-docker-build-push:
    runs-on: ubuntu-latest
    if: github.event.pull_request.merged == true &amp;&amp; ( github.event.pull_request.head.ref == 'develop' || startsWith(github.event.pull_request.head.ref, 'hotfix/') )
    steps:
        ...
</code></pre>
<p>I don’t like the fact of using the branch name in a <code>job</code> condition, but for time purposes, we’ll start with this alternative. I’ll dig into some more API documentation, or Github’s marketplace to found a more reliable way to achieve this :+1:</p>
<p>Thanks again! Cheers!</p>

**Reply by thomasInes** [reply]
Hi, I `github.event.pull_request.merged` is empty when I try to log it, when I merge a PR.
Is it still supported?

**Comment by melink14** [comment]
<p>Don’t feel bad, I think the documentation is a bit lacking when it comes to details and worked examples. (For example, how pull request branch targeting works is not explained in the event trigger docs)</p>
<p>The only real place I could find this was a blurb in the GitHub action syntax docs where it talks about the syntax for adding branch restrictions to <code>pull_request</code> triggers:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" target="_blank" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/workflow-syntax-for-github-actions</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>(The part about the restrictions applying to the <code>base</code>)</p>
<p>Essentially the core of action documentation will come from experience, this forum, and various GitHub repo issue threads. :smiley:</p>
<p>Good luck! Glad you found a short term fix; at first, I thought you were implying that you wanted to push hotfixes directly to the main branch with no PR which would be even harder!</p>
